Fix #1110

RedMammoth opened the pull request:
feat: Добавил разбор настроек прокси из переменных окружения
fix: Исправил basic авторизацию, если useOSAuthentication = false

fix #1110

RedMammoth committed:
feat: Добавил разбор настроек прокси из переменных окружения
fix: Исправил basic авторизацию, если useOSAuthentication = false

fix #1110

EvilBeaver commented on src/ScriptEngine.HostedScript/ScriptEngine.HostedScript.csproj:
```diff
@@ -9,6 +9,7 @@
   <ItemGroup>
     <PackageReference Include="DotNetZip" Version="1.13.3" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="System.Runtime.InteropServices.RuntimeInformation" Version="4.3.0" />
```
А где тут применяется Interop в этой доработке?

RedMammoth replied:
Проверка операционной системы (src/ScriptEngine.HostedScript/Library/Http/InternetProxyContext.cs:39)
```c#
if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
````

EvilBeaver replied:
А это добавит еще одну обязательную DLL в дистрибутив или нет? Если да - то оберните в препроцессор #if NETCORE, а в #else - сделайте по старо-дотнетовски System.Environment.OSVersion

EvilBeaver replied:
У нас даже уже есть метод 
![image](https://user-images.githubusercontent.com/2657561/125930391-f01b1d9c-fafd-42eb-9112-aa7c5233dfd3.png)

RedMammoth committed:
fix: убрал зависимость от InteropServices

Переписал проверку ОС с помощью System.Environment.OSVersion.Platform

RedMammoth committed:
feat: Добавил разбор переменной no_proxy для linux

nixel2007 commented on src/ScriptEngine.HostedScript/Library/Http/InternetProxyContext.cs:
```diff
@@ -36,7 +36,7 @@ public InternetProxyContext(bool useDefault)
             if (useDefault)
             {
-                if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+                if (System.Environment.OSVersion.Platform == System.PlatformID.Unix)
```
для macOS поведение не идентичное?

RedMammoth replied:
На macOS задают переменные для прокси http_proxy, https_proxy, no_proxy? Никогда не сидел, не знаю как там прокси настраивают

nixel2007 replied:
тоже без понятия. тогда оставим до первого маковода с прокси

EvilBeaver replied:
Да, там как в Линуксе. По-крайней мере, последний раз, когда я интересовался вопросом так было.

EvilBeaver commented:
@RedMammoth подниму тему. Реально доделать до прохождения зеленых тестов?

RedMammoth commented:
> @RedMammoth подниму тему. Реально доделать до прохождения зеленых тестов?

Если нужно чтобы текущие тесты прошли, то могу сегодня поправить и сохранять пользователя и пароль для всех типов аутентификации.

Но я хочу поменять тест, чтобы он "честно" проверял, идет ли запрос через прокси. На локальной машине я запустил два контейнера (прокси и веб-сервер, возвращающий IP клиента) и ы тестовой процедуре HTTPОтвет вернул в теле IP прокси-сервера. Осталось положить в jenkins, до него еще не успел добраться. В эти выходные разверну и буду пробовать

EvilBeaver commented:
Будем ждать результатов

EvilBeaver commented:
@RedMammoth получилось посмотреть? Хотелось бы включить в ближайший релиз 1.7

RedMammoth commented:
Развернул локально jenkins, и начал пробовать.

Методика [sidecar](https://www.jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers), которую посоветовал @nixel2007, судя по описанию отлично подходит для выполнения теста с запущенными рядом контейнерами, но эти команды работают только когда Jenkinsfile выполнен в скриптовом стиле. А текущий Jenkinsfile выполнен, если я правильно понимаю, в декларативном.

Затем попробовал "влоб" выполнить команды по запуску контейнеров:
```
 bat "docker network create --subnet=192.168.0.0/24 test_proxy_net"
bat "docker run --net test_proxy_net --ip 192.168.0.9 --name squid_proxy -d --publish 3128:3128 -p 2222:22 -e SQUID_USER=proxy_user -e SQUID_PASS=proxy_pass --volume /var/spool/squid thelebster/docker-squid-simple-proxy"
bat "docker run --net test_proxy_net --ip 192.168.0.5 -d -p 8081:8081 dxxwarlockxxb/test_echo_ip_server"
bat "chcp $outputEnc > nul\r\n\"${tool 'MSBuild'}\" Build.csproj /t:xUnitTest"
```

Такой тест отработал:
```
	Прокси = Новый ИнтернетПрокси();

	Прокси.Установить("http","http://localhost", 3128, "proxy_user", "proxy_pass", Ложь);
	Соединение = Новый HTTPСоединение("http://192.168.0.5", 8081, , , Прокси);
	Ответ = Соединение.Получить(Новый HTTPЗапрос());
	юТест.ПроверитьРавенство(200, Ответ.КодСостояния);

	IP = Ответ.ПолучитьТелоКакСтроку();
	юТест.ПроверитьРавенство("192.168.0.9", IP);
```
Но надо за собой удалять сеть и останавливать контейнеры, или как-то по-другому реализовать (не успел еще прочитать про dind).

В общем еще надо доделать, отсюда вопрос:
1. Сколько у меня еще есть времени до релиза?
2. Я могу попробовать переписать jenkinsfile в скриптовом стиле? или он специально создан в декларативном.

EvilBeaver commented:
Я бы оставил чуть меньшее. Просто убедиться, что код работает, поправить тесты минимально.
То есть прям инфраструктуру поднимать наверное излишне. Или оставть на будущее.

Пара дней до релиза есть.

RedMammoth committed:
feat: изменил тест ПроверитьАвторизациюПрокси

пользователь и пароль используется только при basic авторизации, которая задействуется при установке флага ИспользоватьАутентификациюОС в Ложь

EvilBeaver commented:
У меня нет под рукой сети с Прокси. Работают ли с этими изменениями opm, HttpСоединение?

RedMammoth commented:
HttpСоединение работает с NTML авторизацией под windows, с basic авторизацией под windows, с basic авторизацией под linux. Сейчас еще раз перепроверю и проверю opm

RedMammoth commented:
Скачал инсталляторы с [артефактов](https://build.oscript.io/blue/organizations/jenkins/1Script/detail/PR-1111/5/artifacts/)
На работе сеть с прокси, вроде все работает:
![image](https://user-images.githubusercontent.com/64589646/126982184-627f0dc6-4e11-464e-8853-4d83db4a0254.png)
![image](https://user-images.githubusercontent.com/64589646/126982357-e8a7d5f5-684a-4162-b37d-007c74899999.png)

nixel2007 commented:
> Я могу попробовать переписать jenkinsfile в скриптовом стиле? или он специально создан в декларативном.

В декларативном jenkinsfile есть блок script {}, в котором можно писать произвольный код. Если же нужно явное выделение node, то кложуру node можно использовать внутри stage: https://www.jenkins.io/doc/book/pipeline/syntax/#stage-level-agents
